### ISSUE TITLE
run_as_user for Windows installation

### ISSUE THREAD

**Issue by arash8m** [Esri/arcgis-cookbook#199]
I'm using ArcGIS AMI on AWS that comes pre-installed with ArcGIS products. 
If I  use a specific domain user account for run_as_user parameter, do I need to run the cookbooks for installing ArcGIS products from scratch so that everything is installed under that domain user? 

This is my config:
```
{
  "arcgis" : {
    "version": "10.6.1",
    "run_as_user" : "gccp\\arcgis_svc",
...
```

This is the error when using the pre-installed products:
```
[2019-03-02T14:54:30+00:00] FATAL: RuntimeError: arcgis_enterprise_portal[Create Portal Site] (arcgis-enterprise::portal line 76) had an error: RuntimeError: The user arcptl-6cd5cfd\arcgis d
oes not have full control permission to the directory D:\arcgisportal.
```

It's using the local arcgis user for writing to that folder. 

Thanks.

**Comment by pbobov** [comment]
If you need to use user account other that 'arcgis', you will need to change logon accounts of  windows services of the ArcGIS applications such as 'ArcGISServer', 'Portal for ArcGIS', and 'ArcGIS Data Store'.

:server, :portal and :datastore resources in arcgis-enterprise cookbook provide :update_account actions that are used to update logon accounts of the correspondent windows services. These actions are called by install_* recipes. So far there are no recipes that include just :update_account actions, but you can make one.

**Comment by arash8m** [comment]
Thank you Pavel. I added extra recipes for calling update_account for both Portal and Server and it worked.
The problem I ran into is when bringing up extra Servers and joining them to the site. 
I call update_account for these extra nodes as well and they run the server_node recipe. In the config file, I'm setting use_join_site_tool to true. I'm getting the following error:

`ERROR: Failed to join ArcGIS Server site. The user name or password is incorrect. - CreateProcessWithLogonW`

Would you know if there's anything extra I need to do? The script for joining servers to site was working before I added the changes for setting it up under a different user.

**Comment by arash8m** [comment]
I was able to troubleshoot this further. I believe it has to do something with the format of run_as_user.
The way I had I have it in node.json is:
`arcgis\\gcp.local`

Looking at the join_site function in server.rb, the line that's causing the failure is this:
`Mixlib::ShellOut.new(join_site_tool_cmd, 
            { :user => node['arcgis']['run_as_user'],
              :password => node['arcgis']['run_as_password'],
              :timeout => 1800 })
`

I changed the line to look like this and it was able to run without any problem:
`Mixlib::ShellOut.new(join_site_tool_cmd, 
            { :user => 'arcgis',
             :domain => 'gccp.local',
              :password => node['arcgis']['run_as_password'],
              :timeout => 1800 })
`
So basically separating domain from the user solves the problem. Am I defining the run_as_user in a format that's incorrect?

**Comment by pbobov** [comment]
For some reason chef cannot run join-site.bat tool as using the specified "run_as_user" user account. I suspect that account does not have some privileges on the machine. Please check the value of arcgis/server/use_join_site_tool attribute. Try setting this attribute to false. This will make chef using Admin REST API to join site instead of  join-site.bat tool.

**Comment by pbobov** [comment]
> I was able to troubleshoot this further. I believe it has to do something with the format of run_as_user.
> The way I had I have it in node.json is:
> `arcgis\\gcp.local`
> 
> Looking at the join_site function in server.rb, the line that's causing the failure is this:
> `Mixlib::ShellOut.new(join_site_tool_cmd, { :user => node['arcgis']['run_as_user'], :password => node['arcgis']['run_as_password'], :timeout => 1800 }) `
> 
> I changed the line to look like this and it was able to run without any problem:
> `Mixlib::ShellOut.new(join_site_tool_cmd, { :user => 'arcgis', :domain => 'gccp.local', :password => node['arcgis']['run_as_password'], :timeout => 1800 }) `
> So basically separating domain from the user solves the problem. Am I defining the run_as_user in a format that's incorrect?

You used the correct format for the domain user account. It looks like a bug in :join_site action implementation of the cookbook that is specific for domain accounts and join-site tool.

**Comment by arash8m** [comment]
The servers are in auto-scaling configuration and without the site_join tool it'll use primary_server_url. That URL could be invalid when a new server tries to join if the primary server is destroyed. 
I can't think of any other option aside from just fixing the site_join function based on what I discovered.
Please let me know if you have any better suggestion.

**Comment by arash8m** [comment]
I patched up the site_join function and it's working good, so I'll leave it at that. 
One thing to mention is that there's also a permission issue on the folder where the site_join tool is in. It tries to create a xml config file but it doesn't have permission to write to that folder. (the domain user that's executing the site join tool doesn't have permission to modify that folder content)
